pkg/integrations/v2/eventhandler/eventhandler.go

```diff
@@ -0,0 +1,421 @@
+)
+
+const (
+	cacheFileMode = 0600
+)
+
+// EventHandler watches for Kubernetes Event objects and hands them off to
+// Agent's logs subsystem (embedded promtail).
+type EventHandler struct {
+	LogsClient    *logs.Logs
+	LogsInstance  string
+	Log           log.Logger
+	CachePath     string
+	LastEvent     *ShippedEvents
+	InitEvent     *ShippedEvents
+	EventInformer cache.SharedIndexInformer
+	ClusterLabel  string
+	SendTimeout   time.Duration
+	FlushInterval time.Duration
+	ticker        *time.Ticker
+	sync.Mutex
+}
+
+// ShippedEvents stores a timestamp and map of event ResourceVersions shipped for that timestamp.
+// Used to avoid double-shipping events upon restart.
+type ShippedEvents struct {
+	// shipped event's timestamp
+	Timestamp time.Time `json:"ts"`
+	// map of event RVs (resource versions) already "shipped" (handed off) for this timestamp.
+	// this is to handle the case of a timestamp having multiple events,
+	// which happens quite frequently.
+	RvMap map[string]struct{} `json:"resourceVersion"`
+}
+
+func newEventHandler(l log.Logger, globals integrations.Globals, c *Config) (integrations.Integration, error) {
+	var (
+		config *rest.Config
+		err    error
+	)
+
+	// Try using KubeconfigPath or inClusterConfig
+	config, err = clientcmd.BuildConfigFromFlags("", c.KubeconfigPath)
+	if err != nil {
+		level.Error(l).Log("msg", "Loading from KubeconfigPath or inClusterConfig failed", "err", err)
+		// Trying default home location
+		if home := homedir.HomeDir(); home != "" {
+			kubeconfigPath := filepath.Join(home, ".kube", "config")
+			config, err = clientcmd.BuildConfigFromFlags("", kubeconfigPath)
+			if err != nil {
+				level.Error(l).Log("msg", "Could not load a kubeconfig", "err", err)
+				return nil, err
+			}
+		} else {
+			err = fmt.Errorf("could not load a kubeconfig")
+			return nil, err
+		}
+	}
+
+	clientset, err := kubernetes.NewForConfig(config)
+	if err != nil {
+		return nil, err
+	}
+
+	// get an informer
+	factory := informers.NewSharedInformerFactory(clientset, time.Duration(c.InformerResync)*time.Second)
+	eventInformer := factory.Core().V1().Events().Informer()
+
+	eh := &EventHandler{
+		LogsClient:    globals.Logs,
+		LogsInstance:  c.LogsInstance,
+		Log:           l,
+		CachePath:     c.CachePath,
+		EventInformer: eventInformer,
+		ClusterLabel:  c.ClusterName,
+		SendTimeout:   time.Duration(c.SendTimeout) * time.Second,
+		FlushInterval: time.Duration(c.FlushInterval) * time.Second,
+	}
+	// set the resource handler fns
+	eh.initInformer(eventInformer)
+	return eh, nil
+}
+
+// Initialize informer by setting event handler fns
+func (eh *EventHandler) initInformer(eventsInformer cache.SharedIndexInformer) {
+	eventsInformer.AddEventHandler(cache.ResourceEventHandlerFuncs{
+		AddFunc:    eh.addEvent,
+		UpdateFunc: eh.updateEvent,
+		DeleteFunc: eh.deleteEvent,
+	})
+}
+
+// Handles new event objects
+func (eh *EventHandler) addEvent(obj interface{}) {
+	event, ok := obj.(*v1.Event)
```

Comment:
It's probably to assume throughout the handlers that we'll only receive `*v1.Events` and simplify the code a bit by removing all the `ok` checks